### ISSUE TITLE
Utilizing IP Tagging as an Edge proxy

### ISSUE THREAD

**Issue by cetanu** [envoyproxy/envoy#8412]
I've come across a small issue that I'm unsure how to solve:

When running envoy as an edge proxy, it's advised by the docs to set [`use_remote_address`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#envoy-api-field-config-filter-network-http-connection-manager-v2-httpconnectionmanager-use-remote-address) to `true` in the listener config.

This setting seems to prevent envoy from adding the `X-Envoy-Ip-Tags` header *unless* the IP is a private address from rfc1918.

An example configuration that you may use to trigger such behavior:

```
static_resources:
  listeners:
    - name: test
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 8888
          protocol: TCP
      filter_chains:
        - filters:
            - name: envoy.http_connection_manager
              config:
                stat_prefix: test
                use_remote_address: true
                http_filters:
                  - name: envoy.ip_tagging
                    config:
                      ip_tags: 
                        - ip_tag_name: AWS-EC2-US-EAST-1
                          ip_list:
                            - address_prefix: 52.23.63.224
                              prefix_len: 27
                  - name: envoy.router
                route_config:
                  name: static-response
                  virtual_hosts:
                    - name: catchall
                      domains:
                        - '*'
                      routes:
                        - match:
                            prefix: /
                          direct_response:
                            status: 200
                            body:
                              inline_string: |
                                Hello world!
```

This may be somewhat hard to test unless you can make a request to an envoy proxy running the above config from a public IP address. Maybe if you have a EC2 box in AWS us-east-1 you can test with it? 😅 

Am I perhaps doing something wrong here, or is this a known limitation of the IP Tagging HTTP Filter?
Should I propose a change to the filter?

**Comment by derekargueta** [comment]
Hi @cetanu, the IP-tagging filter, if xff_num_trusted_hops is not configured, is only dependent on the immediate client which is likely your NLB or ELB. If you have an intermediate hop - either due to ELB (not using Proxy protocol and not appending XFF) or a CDN, you may want to configure xff_num_trusted_hops to reflect that. The IP tagging filter depends on the right-most IP returned by x-forwarded-for as mentioned [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ip_tagging_filter#config-http-filters-ip-tagging).

More docs:
[x-forwarded-for](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for) <-- very useful for understanding trusted-client-IP behavior
[xff_num_trusted_hops](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#envoy-api-field-config-filter-network-http-connection-manager-v2-httpconnectionmanager-xff-num-trusted-hops)

**Comment by cetanu** [comment]
We do make use of an NLB, but it's in TCP mode, so there's no modification of the XFF header at that point, therefore it's my understanding that an edge proxy behind a TCP NLB should not make use of `xff_num_trusted_hops`?

Additionally, we did experiment with setting `xff_num_trusted_hops`, but there was no change in behavior until we entirely removed `use_remote_address` (implicitly setting it to false)

**Comment by cetanu** [comment]
As per another section of the docs:

> If use_remote_address is set to true, the request is internal if and only if the request contains no XFF and the immediate downstream node’s connection to Envoy has an internal (RFC1918 or RFC4193) source address

This explains why IP Tagging works when the address is RFC1918, and does not in any other circumstance, unless I set `use_remote_address` to `false`

Unless I am misunderstanding?

**Comment by cetanu** [comment]
### From the envoy documentation

* The HTTP IP Tagging filter sets the header x-envoy-ip-tags with the string tags for the trusted address from x-forwarded-for
* If use_remote_address is set to true, Envoy sets the x-envoy-external-address header to the trusted client address.

### The test we've tried

You can browse to https://edge-echo.services.atlassian.com/

This acts as a lite version of httpbin and will show you what headers it received.

Of course, you won't have the same source IP as me, but this screenshot shows my source address is being added to x-envoy-external-address:

![image](https://user-images.githubusercontent.com/2289018/66616628-55406100-ec1d-11e9-8081-4c40a0ac5d9c.png)

We have an IP tag which contains a CidrRange of: 
```
{
    "address_prefix": "103.233.242.8",
    "prefix_len": "31"
}
```

If I am not mistaken, my source address is within the above CidrRange:

![image](https://user-images.githubusercontent.com/2289018/66616700-9173c180-ec1d-11e9-9bdd-74ba9bd081ed.png)

We have `use_remote_address` set to `True`,
and `xff_num_trusted_hops` set to `0`

The question remains... why isn't the IP Tag filter adding the header, considering that the trusted IP is added to x-envoy-external-address, and it is within one of the CidrRanges configured?

**Comment by derekargueta** [comment]
Is the 10.x.x.x IP in `x-forwarded-for` the internal IP address of your Envoy instance? Assuming you don't have some kind of cloud load balancer in front (the DNS looks like it points straight to ec2) the workflow is
1) you connect directly to the Envoy ec2 instance, Envoy adds your IP to XFF
2) request goes on to your webapp, which adds Envoy's IP to XFF before presenting.
Is this correct?

If there's a cloud load balancer in front that is modifying XFF and the webapp is not modifying XFF, then this whole situation makes sense. Otherwise it's back to the drawing board :/

**Comment by cetanu** [comment]
@derekargueta 

My apologies, I should have explained that in the previous comment. 
This 'echo' page displays what it received from envoy, which means the XFF header there has been appended to by envoy, so that 10/8 address is our proxy.

**Comment by cetanu** [comment]
😰 I haven't had a ton of time to look more into this. Any avenues I should look into would help push me along

**Comment by cetanu** [comment]
After looking at some of the settings, and wondering why setting `use_remote_address` to `false` would cause the Ip-Tagging to suddenly work...

The following seems to be the case, to determine internal/external + manipulate headers, when `use_remote_address` is...

* `false`: envoy uses the xff header
* `true`: envoy uses the real client address

Considering this, could it be the case that when it is `true`, there is no xff header, and given that ip tagging relies on this header, it would not execute correctly? Is this a timing/ordering problem?

**Comment by cetanu** [comment]
Pre-empting stalebot...

I still haven't figured out what's going on here (I haven't had the time though)

but we still really want to know if we're doing something wrong, or if this is a bug

**Comment by cetanu** [comment]
Just checking in again, still haven't figured out why this is going on, but also haven't had that much time to look at it.

**Comment by derekargueta** [comment]
@cetanu sorry this slipped to the wayside. I'm going to try to block off some time this week to get to the bottom of this. I think understanding this scenario is good fodder for improving the documentation around IP tagging and resolving "true client IP".

**Comment by cetanu** [comment]
👋 I have some time to collaborate on this now that the holiday season has passed

**Comment by cetanu** [comment]
Okay so this is weird - we exposed this parameter internally that allowed teams to use ip-tags to 'allowlist' their routes, but with a big warning sign saying "⚠️  Hey this doesn't work yet!"

Anyway, we started rejecting this parameter, and someone chirped up and said "We were using that!"... So I started investigating and I _think_ that it's working for us now.

Here's my theory:

1. request comes in
2. envoy looks up external address in the ip-tags
3. envoy adds `X-Envoy-Ip-Tags` as an `INTERNAL` header
4. the HeaderMatcher in our RouteMatch will successfully find this tag and match on it
5. The header is NOT proxied to the upstream

Is this correct? If so we might be able to close this off 😀
